compiler/rustc_hir_typeck/src/method/suggest.rs

```diff
@@ -2322,7 +2323,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             .iter()
             .map(|e| (e.obligation.predicate, None, Some(e.obligation.cause.clone())))
             .collect();
-        self.suggest_derive(err, &preds);
+        self.suggest_derive(err, &preds, suggest_derive);
     }
 
     pub fn suggest_derive(
```

Comment:
I find it confusing that this method is called `suggest_derive` but apparently does more than suggesting `#[derive]`. It also adds the note *the trait[…] […] must be implemented*. Adding the flag `suggest_derive` to … `suggest_derive` makes it even more confusing. Could this method be renamed or even refactored into two methods potentially allowing us to get rid of the `bool` parameter altogether?

Reply:
I wanted to make this as minimal a diff as possible :)

`suggest_derive` *also* emits some extra notes (that I didn't want to lose). I can either rename the method and keep the argument, or split the logic in two (or three?) methods and call them only when needed.

Reply:
I see :) If you can find a good name for it, go for a rename. You can do the refactor in this PR or or in a follow-up one, larger diffs are fine by me.

Reply:
What do you think of the change?

Reply:
Thanks, it's an improvement. Although, the way it's split into two methods doesn't feel right: `note_predicate_source_and_get_derives` still does two things, it calculates `traits` and `derives` and only adds a note for `traits` but never for `derive` and we do a bunch of work formatting `derives` just to throw it away if `!suggest_derive`.